Forces to thruster signal

joaobrittoneto opened the pull request:
Convert thrusters forces to thruster's command signal, according to its dynamics. 
To be used between AccelerationController and thruster's driver.

joaobrittoneto committed:
convert forces to thruster signal

joaobrittoneto committed:
Include raw (pwm) thruster input

doudou commented on tasks/ThrustersInput.cpp:
```diff
@@ -0,0 +1,192 @@
+    		// Let's set control_modes with default values RAW
+    		numberOfThrusters = coeffPos.size();
+        	//resize the controlModes to number of Thrusters
+        	controlModes.resize(numberOfThrusters);
```
resize() can be provided with the initialization value

```
controlModes.resize(numberOfThrusters, base::JointState::RAW)
```

doudou commented on tasks/ThrustersInput.cpp:
```diff
@@ -0,0 +1,192 @@
+    {
+    	if(controlModes[i] !=  base::JointState::SPEED && controlModes[i] !=  base::JointState::RAW)
+    	{
+			RTT::log(RTT::Error) << "Control Mode should be SPEED or RAW" << RTT::endlog();
```
Wrong indentation, you're mixing tabs and spaces and have your editor configured to do tab == 4 spaces.

Use tab==8 and indent with spaces.

doudou commented on tasks/ThrustersInput.cpp:
```diff
@@ -0,0 +1,192 @@
+    // Check size of parameters. They should met among them.
+    if(coeffPos.size() != numberOfThrusters || coeffNeg.size() != numberOfThrusters )
+    {
+    	if(coeffPos.size() != coeffNeg.size())
```
This makes the whole logic hard to read, and I believe will be hard to understand on the usage side as well. All three have to agree, so test for that and output a generic exception (THRUSTER_ARRAYS_SIZE_DIFFER), message

```
The size of control_modes, thruster_coeff_pos and thruster_coeff_neg do not agree (... output the size of the 3 ...)
```

doudou commented on tasks/ThrustersInput.cpp:
```diff
@@ -0,0 +1,192 @@
+    	if(controlModes[i] !=  base::JointState::SPEED && controlModes[i] !=  base::JointState::RAW)
+    	{
+			RTT::log(RTT::Error) << "Control Mode should be SPEED or RAW" << RTT::endlog();
+		exception(WRONG_SET_OF_CONTROLMODES);
```
return false if there is a configuration problem.

doudou commented on tasks/ThrustersInput.cpp:
```diff
@@ -0,0 +1,192 @@
+    	else if(!controlModes.empty())
+    	{
+    		// Well, if the thruster_coeff vectors have the same size, looks like the control_modes is the one with wrong size
+    		exception(WRONG_SIZE_OF_CONTROLMODES);
```
return false if there is a configuration problem

doudou commented on tasks/ThrustersInput.cpp:
```diff
@@ -0,0 +1,192 @@
+    	if(coeffPos.size() != coeffNeg.size())
+    	{	// If coefficients sizes are different among them and controm_modes size, one of the coeff were not set right
+    		RTT::log(RTT::Error) << "Coefficients were not set with "<< numberOfThrusters <<" as predicted" << RTT::endlog();
+    		exception(WRONG_SIZE_OF_THRUSTER_COEFFICIENTS);
```
return false if there is a configuration problem

doudou commented on tasks/ThrustersInput.cpp:
```diff
@@ -0,0 +1,192 @@
+    if(coeffPos.empty() || coeffNeg.empty())
+    {
+		RTT::log(RTT::Error) << "Thruster coefficients were not set" << RTT::endlog();
+    	exception(WRONG_SIZE_OF_THRUSTER_COEFFICIENTS);
```
return false if there is a configuration problem

doudou commented on tasks/ThrustersInput.cpp:
```diff
@@ -0,0 +1,192 @@
+	for (int i=0; i<numberOfThrusters; i++)
+	{
+		// Define how to call the problematic thruster
+		std::string textThruster;
```
Don't do this computation if you don't need it, i.e. move it to the if (ERROR) { } below

doudou commented on tasks/ThrustersInput.cpp:
```diff
@@ -0,0 +1,192 @@
+	{
+		// Define how to call the problematic thruster
+		std::string textThruster;
+		if(cmd_in.size() == numberOfThrusters)
```
Isn't this always going to be true ? You check for it above ...

Commit:
Corrected the identation and other minor bugs.

Commit:
Corrected configuration error checkings.

rafaelsaback commented:
Corrected the error checking and improved the code as a whole. I also used some unitary tests to check the configuration exceptions and also the calculated values.

rafaelsaback commented:
I've just realized that we still need to implement limits for the output. To be done soon...

Commit:
ThrustersInput: fix "leaking" input command to the output

calcOutput was modifying the input in-place instead of creating
a new structure. Fix that, and fix some tests that were too lax.

rafaelsaback commented:
@doudou, added the missing commit here

Commit:
copied the time field from the input to the output

saarnold commented:
One could also allow efforts and just hand them through, but it's not necessary.
Is this task complete, can it be merged?

rafaelsaback commented:
Added a missing commit here.

@saarnold, I'm positive about this suggestion, do you see any significative drawback here @doudou?

doudou commented:
Yes, I'm strongly against.

If you generate efforts, don't add a component that does nothing with it -- adding complexity on both sides (the component will become more complex than required)
